src/hypergraph.jl

```diff
@@ -295,4 +295,38 @@ function nhv(h::Hypergraph{T, V, E}) where {T <: Real, V, E}
     length(h.v2he)
 end
 
+function _default_heselect(h::Hypergraph, v::Int)
+    hes = gethyperedges(h, v)
+    collect(keys(hes)), ones(length(hes))
+end
+
+function _default_vselect(h::Hypergraph, he::Int)
+    vs = getvertices(h, he)
+    collect(keys(vs)), ones(length(vs))
+end
+
+"""
+    random_walk(h::Hypergraph, start::Int; heselect::Function, vselect::Function)
+
+Return a next vertex visited in assuming a random walk starting from vertex `start`.
+First a hyperedge is sampled with weights proportional to `heselect` function
+(by default each hyperedge is sampled with the same probability).
+Next a vertex within hyperedge is with weights proportional to `vselect` function
+(by default each vertex, including the source, is sampled with the same probability).
+
+`heselect` and `vselect` functions take two arguments a `Hypergraph` and respectively
+a vertex identifier or a hyperedge identifier. The return values of both functions
+should be respectively a list of hyperedges or vertices and their weights.
+"""
+
+function random_walk(h::Hypergraph, start::Int;
```

Comment:
This code is nondeterministic. The order of results is dependent on data layout in the dictionary keys. Hence calling twice `random_walk` with the same `Random.seed!` value might result in different random walks. This unstability makes it inappropriate for numerical experiments.  Rather than that, the `random_walk` function should perhaps built a sorted cache of vertices and hyperedges and iterate over such cache. Perhaps it might make sense to introduce something like `SortedEdges(h::Hypegrah)` ? or simply built-in this sorted cache inside the `random_walk` function.

Reply:
It is deterministic. You will only have a difference between 32-bit and 64-bit Julia. But you do not currently test this package against 32-bit Julia (do we want to support this portability - I am OK to do so if you think it is worth it).

In general - if we think that order of iteration over hyperedges/vertices in SimpleHypergraphs.jl is important (here I see the value in it) maybe let us switch `Dict` to `OrderedDict` (this will have probably ~20% performance hit) or allow the user to choose the collection one wants to use for internal storage upon construction of `Hypergraph` object.

Reply:
It is creation-order deterministic but not deterministic for two identical hypergraphs. That is if we have situation where `h1 == h2` evaluates to `true` and `all(Matrix(h1) .=== Matrix(h2))` also evaluates to `true`  and still identical random seeds might lead to different random walks. 

I mean here a behavior such as this one (just for simplicity I focus on dictionaries):
```
julia> d1 = Dict(1:30 .=> 101:130); d2 = Dict(30:-1:1 .=> 130:-1:101);
julia> d1 == d2
true
julia> collect(keys(d1)) == collect(keys(d2))
false
```

Or, even more surprising, consider this code:
```
julia> using Serialization
julia> serialize("d2.o",d2)
julia> d3 = deserialize("d2.o");
julia> collect(keys(d2))==collect(keys(d3))
false
```
In practice this means very big trouble for using common random numbers in any multiprocessing environment. 

The key order in an `OrderedDict` is also insertion order dependent so it would not solve the issue. 

Of course it is a design decision to leave it that way but in future users might have errors difficult to debug.

Reply:
I know, I just wanted to have something hard-coded in `random_walk` as this should be solved on `Hypergraph` level as other algorithms will hit similar problems (maybe even the algorithms we already have). We might use `SortedDict` which also should be OK (we do not aim for creation speed anyway, as mostly we are interested in static hypergraphs for now).
Anyway - I will add sorting to `random_walk` for now.

Reply:
Yes `SortedDict` would solve the issue. 
Maybe we could add the underlying dict type to the Hypergraph constructor? In this way the user could decide.

Reply:
This is what I have suggested earlier (to allow the choice of the container, e.g. some users might also want `RobinDict` in special cases). Just make a `Dict` the default.

Reply:
This will make a seperate PR, but since the idea is inspired here, I propose the following `Hypergraph` constructor:
```julia
struct Hypergraph{T<:Real, D<:AbstractDict{Int,T},V,E} <: AbstractMatrix{Union{T, Nothing}} 
      v2he::Vector{D}                                         
      he2v::Vector{D}                                     
      v_meta::Vector{Union{V,Nothing}}                          
      he_meta::Vector{Union{E,Nothing}}                     
      Hypergraph{T,D,V,E}(n, k,                           
              v_meta=Vector{Union{V,Nothing}}(nothing, n),              
              he_meta=Vector{Union{E,Nothing}}(nothing, k)              
              ) where {T<:Real,D<:AbstractDict{Int,T}, V, E} =                                 
          new{T,D,V,E}(                                                   
              [D() for i in 1:n],[D() for i in 1:k],
              v_meta, he_meta)                              
end  
```
If it initially looks OK, I can prepare a PR for this

Reply:
yes - with the default for `D` to be `Dict`.